### ISSUE TITLE
Canonical 3D orientation

### ISSUE THREAD

**Issue by romainVala** [MIC-DKFZ/nnUNet#2768]
I notice that during inference, there is no reordering to canonical orientation (as propose by nibable with ` nib.as_closest_canonical() `function 
I found a confirmation in this post  by @TaWald in [#2400](https://github.com/MIC-DKFZ/nnUNet/issues/2400#issuecomment-2459042451)_

predicting a segmentation from a "coronal" oriented volume give me very wrong example whereas reodering the voxel order (with as_closet_canonical) solve the issue

First I was surprised, to see so much different because I though than nnunet augmentation include so kind of axis swaping. (but is it only changing the sign of a given axis, and not changing the voxel ordering (ie from axial to sagital) ? can you confirme what are the exact augmentation used by default

Having the option to apply the as_closet_coanonical() before running the prediction will be a great use to avoid very bad result just because of a strange voxel ordering ... (which is hard to figure out since those "badly" oriented images, just looks fine in any viewer reading correctly the nifti header)

For now I have to handle it myself, (read the input data, reorder it and write it back to disque before calling nnunet prediction) which is quite ineficient compared to have on the fly

**Comment by vcasellesb** [comment]
Hi! I also stumbled across this "bug" -- I don't think it could be considered as such --, when loading the preprocessed npz files, i.e., the orientation of the ndarrays differed between "training cases". This makes sense, since no spatially-aware (i.e., that take into account the header/affine of the image) transformations such as transpositions or rotations are done, besides resampling (scaling) and bringing the anisotropic dimension to the front (X dimension) of the array. 

To implicitly bring all images to the closest and easiest approximation of a common orientation without using a template, I have used SimpleITK to reorient all cases on load during preprocessing in [my fork](https://github.com/vcasellesb/nnUNetv2/blob/ec4d62879dabbe849caf5425cfe51a908408334f/nnunetv2/imageio/simpleitk_reader_writer.py#L138). 


Could this be of interest to you @romainVala ?

**Comment by romainVala** [comment]
Hello
Yes I agree this is not a bug, it would just be more convenient
thanks for pointing your PR. Yes this is what I need.  I hesitate to use a fork of nnunet, but I will try to isolate just this part. (For now I use torchio to reorient, but then I need to save the new oriented volume on disk, and remove it at the end) 

Having  an extra argument to force the voxel ordering of  the input volume directly in the  nnUNetPredictor class, would avoid some big loss in performance just because of an voxel ordering issue ... if I could convince @FabianIsensee  ....

**Comment by Karol-G** [comment]
Hey,

Is this issue still relevant? Otherwise, I will close it in 14 days.

Best,
Karol

**Comment by vcasellesb** [comment]
Hi @Karol-G 

I believe it can be closed, at least on my part.

Best,
Vicent

**Comment by TaWald** [comment]
FYI @vcasellesb and @romainVala :  If you override the reader writer class to`NibabelIOWithReorient` after planning, then preprocessing will create the data in a homogeneous orientation for model training. During inference, nnU-Net should use the same reader-writer class so it should carry over into the prediction without any hassle.

**Comment by vcasellesb** [comment]
> FYI [@vcasellesb](https://github.com/vcasellesb) and [@romainVala](https://github.com/romainVala) : If you override the reader writer class to`NibabelIOWithReorient` after planning, then preprocessing will create the data in a homogeneous orientation for model training. During inference, nnU-Net should use the same reader-writer class so it should carry over into the prediction without any hassle.

True, thanks! But don't discrepancies in how nibabel and SimpleITK load arrays mess up how the network then expects volumes to be oriented (as in, first axis sagittal, second one coronal, third one axial and viceversa)? Or the code already handles that?

**Comment by vcasellesb** [comment]
> > FYI [@vcasellesb](https://github.com/vcasellesb) and [@romainVala](https://github.com/romainVala) : If you override the reader writer class to`NibabelIOWithReorient` after planning, then preprocessing will create the data in a homogeneous orientation for model training. During inference, nnU-Net should use the same reader-writer class so it should carry over into the prediction without any hassle.
> 
> True, thanks! But don't discrepancies in how nibabel and SimpleITK load arrays mess up how the network then expects volumes to be oriented (as in, first axis sagittal, second one coronal, third one axial and viceversa)? Or the code already handles that?

Yes, I can see that NibabelIO [indeed already handles this](https://github.com/MIC-DKFZ/nnUNet/blob/2932ced87970a288be4edd99b0fac29904fcab9e/nnunetv2/imageio/nibabel_reader_writer.py#L136).

**Comment by TaWald** [comment]
The `NibabelIOWithReorient` depends on the metadata of the nifti to be correct. As long as this is accurate, all the images should be reoriented to the same orientation, which the model is then trained on. So in theory there should not be any issues. If you change the reader in the plans after training, it will indeed lead to issues. Also, if your metadata is wrong, your model will struggle since it's expecting the orientation in a different way.

A second factor in this story is the augmentations, which depend on whether your data is isotropic or anisotropic. If it's Isotropic augmentations include rotation and mirrors along the isotropic axis, which should lead to learned invariances in how the image is reoriented. If you are anisotropic, the incorrect orientations however are much more problematic.

Also don't confuse `NibabelIOWithReorient` with `NibabelIO` which are two different entities with the former doing the reorientation

**Comment by vcasellesb** [comment]
True, I was asking exclusively about discrepancies arising from a change in loader backend. What my question is: assuming the file metadata is correct, and that all images are in the same arbitrary reorientation, simply changing from SimpleITKIO to NibabelIO, regardless of when you do said change (if it's after training or after planning) should not cause any issues (unless the change you are making is from an IO that does reorientation to one that does not, or viceversa -- but that happens even without changing backend).

Sorry If I'm being too obtuse.

**Comment by TaWald** [comment]
Are you saying that swapping the readers after training (to one with the same expected orientation) breaks the predictive behavior? 
That shouldn't actually happen - which readers are you using?
Independent of this, I'm not sure I understand the use case for swapping readers post-hoc. 
Is there any motivation for this? If so one could look into creating a fix

**Comment by vcasellesb** [comment]
> Are you saying that swapping the readers after training (to one with the same expected orientation) breaks the predictive behavior? That shouldn't actually happen - which readers are you using? Independent of this, I'm not sure I understand the use case for swapping readers post-hoc. Is there any motivation for this? If so one could look into creating a fix

No, no, I was just asking to clarify the behavior of both backends, not because I am having an issue. Don't worry about it. Thanks so much for your support.

**Comment by romainVala** [comment]
Hi
I did not have time to test the workaround, but reading the exchange I am not sure it will work
may be I can clarify
In my use case I already trained the model with native nnunet reader, and my concern
was for prediction only, where data is the wild have many different orientation, and I want the canonical one
but in the same "internal state" as the original nnunet 

"internal state" means the 3D matrix data ordering, which is different between nibabel and stik (different for the same nifti volume)

**Comment by vcasellesb** [comment]
Hi @romainVala,

I am no expert, but if I can offer my two-cents: the inference or "data in the wild" as you mention should match the training data – in terms of intensity distribution, spatial properties (spacing, orientation) – as much as humanly possible.

For example, if you trained on a dataset of MRI scans that have all been registered to MNI space, then inference should, at the very least, perform a registration to said space.

Thus, I would say: It does not matter what reader class you used during training, but you should take into account that however the network saw the data during training should be matched during inference. If all your data was in RAS orientation (first axis goes left->right, second posterior->anterior, third inferior->superior), then either you can change the reader writer to handle that, or do that separately prior to feeding the data to nnUNet.

Note that reorienting with nibabel or SimpleITK does not make the axes perfectly align with the grid, it only changes the order of the dimensions in the numpy array and optionally flips the direction. But for more sophisticated stuff (e.g., to achieve orthogonal axes) you'd have to register or resample (interpolate) to a standard reference space.

Can you describe the properties of the data your model was trained on?

**Comment by romainVala** [comment]
I trained with (brain) synthetic data, generated with random contrast. Therefore I obtain a very robust model which is contrast agnostic (and yes I can really predict from any MRI contrast, even a CT scan !) 
(have a look [here](https://github.com/romainVala/SIAM) you'll find more detail in the preprint)

my input data were in canonical orientation, and I used only small affine/rotatin as augmentation during the training

Therefor at inference I need to get close enough to this canonical orientation

I already have a working solution it is just very subotimal ...  I use [torchio](https://docs.torchio.org/transforms/preprocessing/ToCanonical/) to convert the input to canonical space, write it to disk, then run nn-prediction)

**Comment by romainVala** [comment]
> If all your data was in RAS orientation (first axis goes left->right, second posterior->anterior, third inferior->superior), then either you can change the reader writer to handle that, or do that separately prior to feeding the data to nnUNet

ok may be changing the reader would work here  (but doing it separately is  is exactly what I was trying to avoid ...

**Comment by vcasellesb** [comment]
Yes, I've seen your preprint, looks cool. You use SynthSeg, I've also used it :)

What I'd do is change the reader/writer class to the one [I referenced before](https://github.com/vcasellesb/nnUNetv2/blob/ec4d62879dabbe849caf5425cfe51a908408334f/nnunetv2/imageio/simpleitk_reader_writer.py#L138). Now that I see it, a similar solution has already been implemented in the [official nnUNet codebase](https://github.com/MIC-DKFZ/nnUNet/blob/2932ced87970a288be4edd99b0fac29904fcab9e/nnunetv2/imageio/simpleitk_reader_writer.py#L132), that achieves the same. So you don't have to use my fork!

You have to make sure that this doesn't break anything nnUNet expects downstream (the preprocessor, etc.). From what I see on their `simpleitk_reader_writer.py` file, I think that there should be no issues, but please note that I am not an nnUNet developer and my opinion counts as much as any random person online.

The other option I see is using a custom dataloader that implements your desired logic (reading as canonical), and then feed it to the [nnUNet inference method that expects arrays](https://github.com/MIC-DKFZ/nnUNet/blob/2932ced87970a288be4edd99b0fac29904fcab9e/nnunetv2/inference/predict_from_raw_data.py#L453). Please note their method description, as it is highly relevant:

```
input_image: Make sure to load the image in the way nnU-Net expects! nnU-Net is trained on a certain axis
                     ordering which cannot be disturbed in inference,
                     otherwise you will get bad results. The easiest way to achieve that is to use the same I/O class
                     for loading images as was used during nnU-Net preprocessing! You can find that class in your
                     plans.json file under the key "image_reader_writer". If you decide to freestyle, know that the
                     default axis ordering for medical images is the one from SimpleITK. If you load with nibabel,
                     you need to transpose your axes AND your spacing from [x,y,z] to [z,y,x]!
```

**Comment by romainVala** [comment]
(I did not used SynthSeg, just the same method (synthetic training) )

thanks for pointing the new functionality of the official nnunet codebase, that's seems indeed to address this issue ... (not sure why @TaWald  did not mention it ... this would indeed solve the issue, which is any way close now )
